src/libUtils/Logger.cpp

```
@@ -16,325 +16,264 @@
+    m_logRotate.setMaxLogSize(max_file_size_in_bytes);
+  }
+
+  void setMaxArchiveLogCount(int max_size) {
+    m_logRotate.setMaxArchiveLogCount(max_size);
+  }
+
+  void receiveLogMessage(LogMessageMover logEntry) {
+    logMessageCommon(m_stream, logEntry.get());
+    m_logRotate.save(m_stream.str());
+    m_stream.str("");
+  }
+
+ protected:
+  std::ostringstream m_stream;
+  LogRotate m_logRotate;
+};
+
+class GeneralLogSink : public CustomLogRotate {
+ public:
+  using CustomLogRotate::CustomLogRotate;
+
+  void receiveLogMessage(LogMessageMover logEntry) {
+    m_stream << Level(logEntry.get());
+    logMessageCommon(m_stream, logEntry.get());
+    m_logRotate.save(m_stream.str());
+    m_stream.str("");
+  }
+};
+
+class StateLogSink : public CustomLogRotate {
+ public:
+  using CustomLogRotate::CustomLogRotate;
+};
+
+class EpochInfoLogSink : public CustomLogRotate {
+ public:
+  using CustomLogRotate::CustomLogRotate;
+};
+
+class StdoutSink {
+ public:
+  void forwardLogToStdout(LogMessageMover logEntry) {
+    logMessageCommon(m_stream, logEntry.get());
+    std::cout << m_stream.str();
+    m_stream.str("");
+  }
 
-Logger::Logger(const char* prefix, bool log_to_file, const char* logpath,
-               streampos max_file_size) {
-  this->m_logToFile = log_to_file;
-  this->m_maxFileSize = max_file_size;
-  this->m_logPath = logpath;
+ private:
+  std::ostringstream m_stream;
+};
 
+template <typename LogRotateSinkT>
+void AddFileSink(LogWorker& logWorker, const std::string& filePrefix,
+                 const boost::filesystem::path& filePath,
+                 int maxFileSize /*= MAX_FILE_SIZE*/) {
+  auto logFileRoot = boost::filesystem::absolute(filePath);
+  bool useDefaultLocation = false;
   try {
-    if (!boost::filesystem::create_directory(this->m_logPath)) {
-      if ((boost::filesystem::status(this->m_logPath).permissions() &
+    if (!boost::filesystem::create_directory(logFileRoot)) {
```

The standard patter (pre-boost) is to just do an mkdir, then bail out with an error if that breaks, just interpreting the errno.

No need to check permissions etc. - just fail loudly.  Makes sense to just crash if things are misconfigured.